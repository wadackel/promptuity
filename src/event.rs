//! A module to read events.
//!
//! # Event
//!
//! The `event` module provides functionalities used in prompt handlers.
//! The main offering is the direct re-export of [`crossterm::event`].
//!
//! ## Examples
//!
//! ```ignore
//! use promptuity::event::{KeyCode, KeyModifiers};
//!
//! impl<W: std::io::Write> Prompt<W> for YourPrompt {
//!   // ...
//!   fn handle(&mut self, code: KeyCode, modifiers: KeyModifiers) -> PromptState {
//!     match (code, modifiers) {
//!       (KeyCode::Enter, KeyModifiers::NONE) => PromptState::Submit,
//!       _ => PromptState::Active,
//!     }
//!   }
//!   // ...
//! }
//! ```

/// [`KeyCode`] re-exports from [`crossterm::event`].
pub use crossterm::event::KeyCode;
/// [`KeyModifiers`] re-exports from [`crossterm::event`].
pub use crossterm::event::KeyModifiers;
