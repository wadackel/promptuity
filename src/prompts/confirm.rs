use crate::event::*;
use crate::style::{Color, Styled, Symbol};
use crate::{Prompt, PromptInput, PromptState, RenderPayload};

const S_ACTIVE: Symbol = Symbol("●", ">");
const S_INACTIVE: Symbol = Symbol("○", " ");

/// A trait for formatting the [`Confirm`] prompt.
///
/// All methods have default implementations, allowing you to override only the specific formatting process you need.
///
/// # Examples
///
/// ```no_run
/// use promptuity::prompts::{Confirm, ConfirmFormatter};
///
/// struct CustomFormatter;
///
/// impl ConfirmFormatter for CustomFormatter {
///     fn layout(&self, yes: String, no: String) -> String {
///         format!("{} or {}", yes, no)
///     }
/// }
///
/// let _ = Confirm::new("...").with_formatter(CustomFormatter);
/// ```
pub trait ConfirmFormatter {
    /// Formats the "Yes" option.
    fn yes(&self, active: bool) -> String {
        let icon = if active {
            Styled::new(S_ACTIVE).fg(Color::Green).to_string()
        } else {
            Styled::new(S_INACTIVE).fg(Color::DarkGrey).to_string()
        };
        format!("{} Yes", icon)
    }

    /// Formats the "No" option.
    fn no(&self, active: bool) -> String {
        let icon = if active {
            Styled::new(S_ACTIVE).fg(Color::Green).to_string()
        } else {
            Styled::new(S_INACTIVE).fg(Color::DarkGrey).to_string()
        };
        format!("{} No", icon)
    }

    /// Formats the layout of the active prompt.
    fn layout(&self, yes: String, no: String) -> String {
        format!("{}  /  {}", yes, no)
    }

    /// Formats the submitted value.
    fn submit(&self, value: bool) -> String {
        if value {
            "Yes".into()
        } else {
            "No".into()
        }
    }
}

/// The default formatter for [`Confirm`].
pub struct DefaultConfirmFormatter;

impl DefaultConfirmFormatter {
    #[allow(clippy::new_without_default)]
    pub fn new() -> Self {
        Self
    }
}

impl ConfirmFormatter for DefaultConfirmFormatter {}

/// A prompt for inputting a Yes/No choice.
///
/// # Options
///
/// - **Formatter**: Customizes the prompt display. See [`ConfirmFormatter`].
/// - **Hint**: A message to assist with field input. Defaults to `None`.
/// - **Default Value**: The default value of `bool`. Defaults to `false`.
///
/// # Examples
///
/// ```no_run
/// use promptuity::prompts::Confirm;
///
/// let _ = Confirm::new("Are you sure?").with_default(true);
/// ```
pub struct Confirm {
    formatter: Box<dyn ConfirmFormatter>,
    message: String,
    hint: Option<String>,
    value: bool,
}

impl Confirm {
    /// Creates a new [`Confirm`] prompt.
    pub fn new(message: impl std::fmt::Display) -> Self {
        Self {
            formatter: Box::new(DefaultConfirmFormatter),
            message: message.to_string(),
            hint: None,
            value: false,
        }
    }

    /// Sets the formatter for the prompt.
    pub fn with_formatter(&mut self, formatter: impl ConfirmFormatter + 'static) -> &mut Self {
        self.formatter = Box::new(formatter);
        self
    }

    /// Sets the hint message for the prompt.
    pub fn with_hint(&mut self, hint: impl std::fmt::Display) -> &mut Self {
        self.hint = Some(hint.to_string());
        self
    }

    /// Sets the default value for the prompt.
    pub fn with_default(&mut self, value: bool) -> &mut Self {
        self.value = value;
        self
    }
}

impl AsMut<Confirm> for Confirm {
    fn as_mut(&mut self) -> &mut Self {
        self
    }
}

impl Prompt for Confirm {
    type Output = bool;

    fn handle(&mut self, code: KeyCode, modifiers: KeyModifiers) -> PromptState {
        match (code, modifiers) {
            (KeyCode::Enter, _) => PromptState::Submit,
            (KeyCode::Esc, _) | (KeyCode::Char('c'), KeyModifiers::CONTROL) => PromptState::Cancel,
            (KeyCode::Char('y'), KeyModifiers::NONE) | (KeyCode::Char('Y'), KeyModifiers::NONE) => {
                self.value = true;
                PromptState::Submit
            }
            (KeyCode::Char('n'), KeyModifiers::NONE) | (KeyCode::Char('N'), KeyModifiers::NONE) => {
                self.value = false;
                PromptState::Submit
            }
            (KeyCode::Left, _)
            | (KeyCode::Char('h'), _)
            | (KeyCode::Char('p'), KeyModifiers::CONTROL) => {
                self.value = true;
                PromptState::Active
            }
            (KeyCode::Right, _)
            | (KeyCode::Char('l'), _)
            | (KeyCode::Char('n'), KeyModifiers::CONTROL) => {
                self.value = false;
                PromptState::Active
            }
            _ => PromptState::Active,
        }
    }

    fn render(&mut self, state: &PromptState) -> Result<RenderPayload, String> {
        let payload = RenderPayload::new(self.message.clone(), self.hint.clone(), None);

        match state {
            PromptState::Submit => {
                Ok(payload.input(PromptInput::Raw(self.formatter.submit(self.value))))
            }

            PromptState::Cancel => Ok(payload),

            _ => Ok(payload.input(PromptInput::Raw(self.formatter.layout(
                self.formatter.yes(self.value),
                self.formatter.no(!self.value),
            )))),
        }
    }

    fn submit(&mut self) -> Self::Output {
        self.value
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_prompt;

    test_prompt!(
        test_hint,
        Confirm::new("test message").with_hint("hint message"),
        vec![]
    );

    test_prompt!(
        test_default_enter,
        Confirm::new("test message").as_mut(),
        vec![(KeyCode::Enter, KeyModifiers::NONE)]
    );

    test_prompt!(
        test_default_yes,
        Confirm::new("test message").with_default(true),
        vec![(KeyCode::Enter, KeyModifiers::NONE)]
    );

    test_prompt!(
        test_move,
        Confirm::new("test message").as_mut(),
        vec![
            (KeyCode::Left, KeyModifiers::NONE),
            (KeyCode::Right, KeyModifiers::NONE),
            (KeyCode::Char('h'), KeyModifiers::NONE),
            (KeyCode::Char('l'), KeyModifiers::NONE),
            (KeyCode::Char('h'), KeyModifiers::CONTROL),
            (KeyCode::Char('l'), KeyModifiers::CONTROL),
            (KeyCode::Char('p'), KeyModifiers::CONTROL),
            (KeyCode::Char('n'), KeyModifiers::CONTROL),
            (KeyCode::Enter, KeyModifiers::NONE),
        ]
    );

    test_prompt!(
        test_direct_lower_yes,
        Confirm::new("test message").with_default(false),
        vec![(KeyCode::Char('y'), KeyModifiers::NONE)]
    );

    test_prompt!(
        test_direct_upper_yes,
        Confirm::new("test message").with_default(false),
        vec![(KeyCode::Char('Y'), KeyModifiers::NONE)]
    );

    test_prompt!(
        test_direct_lower_no,
        Confirm::new("test message").with_default(true),
        vec![(KeyCode::Char('n'), KeyModifiers::NONE)]
    );

    test_prompt!(
        test_direct_upper_no,
        Confirm::new("test message").with_default(true),
        vec![(KeyCode::Char('N'), KeyModifiers::NONE)]
    );
}
