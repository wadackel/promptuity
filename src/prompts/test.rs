use crate::event::*;
use crate::{Prompt, PromptBody, PromptInput, PromptState, RenderPayload};

pub fn render_payload_display(payload: RenderPayload, full: bool) -> String {
    let mut segments = Vec::new();

    if full {
        segments.push(format!("message: {}", payload.message));
        segments.push(format!("hint: {}", payload.hint.unwrap_or("none".into())));
        segments.push(format!(
            "placeholder: {}",
            payload.placeholder.unwrap_or("none".into())
        ));
    }

    segments.push(match payload.input {
        PromptInput::Raw(raw) => format!("input(raw): {}", raw),
        PromptInput::Cursor(cursor) => {
            let (left, cursor, right) = cursor.split();
            format!("input(cursor): {}|{}{}", left, cursor, right)
        }
        PromptInput::None => "input(none):".into(),
    });

    segments.push(match payload.body {
        PromptBody::Raw(raw) => format!("body(raw):\n{}", raw),
        PromptBody::None => "body(none):".into(),
    });

    segments.join("\n")
}

pub fn render_display<T, W>(state: PromptState, prompt: &mut T, full: bool) -> String
where
    T: Prompt<W>,
    W: std::io::Write,
{
    format!(
        "state: {}\n{}",
        state,
        render_payload_display(prompt.render(&state).unwrap(), full)
    )
}

pub fn handle_actions<T, W>(prompt: &mut T, actions: Vec<(KeyCode, KeyModifiers)>) -> String
where
    T: Prompt<W> + Prompt<Vec<u8>>,
    W: std::io::Write,
{
    let mut output = Vec::new();

    output.push(render_display::<T, Vec<_>>(
        PromptState::Active,
        prompt,
        true,
    ));

    for (code, modifiers) in actions {
        let state = Prompt::<Vec<_>>::handle(prompt, code, modifiers);
        let state = match state {
            PromptState::Submit => {
                if let Err(msg) = Prompt::<Vec<_>>::validate(prompt) {
                    PromptState::Error(msg)
                } else {
                    PromptState::Submit
                }
            }
            state => state,
        };
        output.push(render_display::<T, Vec<_>>(state, prompt, false));
    }

    output.join("\n---\n")
}

#[allow(clippy::crate_in_macro_def)]
#[macro_export]
macro_rules! test_prompt {
    ($name: ident, $ty: ty, $prompt: expr, $actions: expr) => {
        #[test]
        fn $name() {
            let output = crate::prompts::test::handle_actions::<$ty, Vec<_>>($prompt, $actions);
            insta::with_settings!({ omit_expression => true }, {
                insta::assert_snapshot!(output);
            });
        }
    };
}
