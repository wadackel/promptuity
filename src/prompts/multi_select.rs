use crate::event::*;
use crate::pagination::paginate;
use crate::prompts::{DefaultSelectFormatter, SelectFormatter};
use crate::{Error, Prompt, PromptBody, PromptInput, PromptState, RenderPayload};

/// A struct representing an option in the [`MultiSelect`] prompt.
#[derive(Debug, Clone)]
pub struct MultiSelectOption<T: Default + Clone> {
    /// The label of the option.
    pub label: String,
    /// The value of the option.
    pub value: T,
    /// The hint of the option.
    pub hint: Option<String>,
    /// The selected flag of the option.
    pub selected: bool,
}

impl<T: Default + Clone> MultiSelectOption<T> {
    /// Creates a new [`MultiSelectOption`] with the given label and value.
    pub fn new(label: impl std::fmt::Display, value: T) -> Self {
        Self {
            label: label.to_string(),
            value,
            hint: None,
            selected: false,
        }
    }

    /// Sets the hint message for the option.
    pub fn with_hint(mut self, hint: impl std::fmt::Display) -> Self {
        self.hint = Some(hint.to_string());
        self
    }
}

/// A trait for formatting the [`MultiSelect`] prompt.
///
/// `MultiSelectFormatter` does not default-implement some of the formatting processes in the trait.  
/// In the implementation of [`DefaultMultiSelectFormatter`], it internally uses [`DefaultSelectFormatter`] for formatting. Please refer to this when implementing custom formats.
///
/// # Examples
///
/// ```no_run
/// use promptuity::prompts::{MultiSelect, MultiSelectOption, MultiSelectFormatter};
///
/// struct CustomFormatter;
///
/// impl MultiSelectFormatter for CustomFormatter {
///     fn option_icon(&self, active: bool, selected: bool) -> String {
///        if selected {
///            "[x]".into()
///        } else if active {
///            " > ".into()
///        } else {
///            "[ ]".into()
///        }
///     }
///
///     fn option_label(&self, label: String, _active: bool, _selected: bool) -> String {
///         label.clone()
///     }
///
///     fn option_hint(&self, hint: Option<String>, active: bool, _selected: bool) -> String {
///         if active {
///             hint.as_ref().map_or_else(String::new, |hint| {
///                 format!(" - {}", hint)
///             })
///         } else {
///             String::new()
///         }
///     }
///
///     fn option(&self, icon: String, label: String, hint: String, _active: bool, _selected: bool) -> String {
///         format!("{} {}{}", icon, label, hint)
///     }
/// }
///
/// let _ = MultiSelect::new("...", vec![MultiSelectOption::new("...", "...")]).with_formatter(CustomFormatter);
/// ```
pub trait MultiSelectFormatter {
    /// Icons displayed for each option.
    fn option_icon(&self, active: bool, selected: bool) -> String;
    /// Formats the label of the option.
    fn option_label(&self, label: String, active: bool, selected: bool) -> String;
    /// Formats the hint message of the option.
    fn option_hint(&self, hint: Option<String>, active: bool, selected: bool) -> String;
    /// Formats the option.
    fn option(
        &self,
        icon: String,
        label: String,
        hint: String,
        active: bool,
        selected: bool,
    ) -> String;

    /// Formats the submitted value.
    fn submit(&self, labels: Vec<String>) -> String {
        labels.join(", ")
    }

    /// Formats the error message for the required flag.
    fn err_required(&self) -> String {
        "This field is required.".into()
    }

    /// Formats errors for selections below the minimum number.
    fn err_min(&self, min: usize) -> String {
        format!("Please select at least {} options.", min)
    }

    /// Formats errors for selections above the maximum number.
    fn err_max(&self, max: usize) -> String {
        format!("Please select no more than {} options.", max)
    }
}

/// The default formatter for [`MultiSelect`].
pub struct DefaultMultiSelectFormatter {
    inner: DefaultSelectFormatter,
}

impl DefaultMultiSelectFormatter {
    #[allow(clippy::new_without_default)]
    pub fn new() -> Self {
        Self {
            inner: DefaultSelectFormatter::new(),
        }
    }
}

impl MultiSelectFormatter for DefaultMultiSelectFormatter {
    fn option_icon(&self, _active: bool, selected: bool) -> String {
        self.inner.option_icon(selected)
    }

    fn option_label(&self, label: String, active: bool, _selected: bool) -> String {
        self.inner.option_label(label, active)
    }

    fn option_hint(&self, hint: Option<String>, active: bool, _selected: bool) -> String {
        self.inner.option_hint(hint, active)
    }

    fn option(
        &self,
        icon: String,
        label: String,
        hint: String,
        active: bool,
        _selected: bool,
    ) -> String {
        self.inner.option(icon, label, hint, active)
    }
}

/// A prompt for selecting multiple elements from a list of options.
///
/// # Options
///
/// - **Formatter**: Customizes the prompt display. See [`MultiSelectFormatter`].
/// - **Hint**: A message to assist with field input. Defaults to `None`.
/// - **Required**: A flag indicating whether to allow no input.
/// - **Minimum Selections**: The minimum number of selections required. Defaults to `0`.
/// - **Maximum Selections**: The maximum number of selections allowed. Defaults to `usize::MAX`.
/// - **Page Size**: The total number of options to displayed per page, used for pagination. Defaults to `8`.
/// - **Validator**: A function to validate the value at the time of submission.
///
/// # Examples
///
/// ```no_run
/// use promptuity::prompts::{MultiSelect, MultiSelectOption};
///
/// let _ = MultiSelect::new("What is your favorite color?", vec![
///     MultiSelectOption::new("Red", "#ff0000"),
///     MultiSelectOption::new("Green", "#00ff00").with_hint("recommended"),
///     MultiSelectOption::new("Blue", "#0000ff"),
/// ]).with_page_size(5);
/// ```
pub struct MultiSelect<T: Default + Clone> {
    formatter: Box<dyn MultiSelectFormatter>,
    message: String,
    hint: Option<String>,
    required: bool,
    min: usize,
    max: usize,
    page_size: usize,
    options: Vec<MultiSelectOption<T>>,
    index: usize,
}

impl<T: Default + Clone> MultiSelect<T> {
    /// Creates a new [`MultiSelect`] prompt with the given message and options.
    pub fn new(message: impl std::fmt::Display, options: Vec<MultiSelectOption<T>>) -> Self {
        Self {
            formatter: Box::new(DefaultMultiSelectFormatter::new()),
            message: message.to_string(),
            hint: None,
            required: true,
            min: 0,
            max: usize::MAX,
            page_size: 8,
            options,
            index: 0,
        }
    }

    /// Sets the formatter for the prompt.
    pub fn with_formatter(&mut self, formatter: impl MultiSelectFormatter + 'static) -> &mut Self {
        self.formatter = Box::new(formatter);
        self
    }

    /// Sets the hint message for the prompt.
    pub fn with_hint(&mut self, hint: impl std::fmt::Display) -> &mut Self {
        self.hint = Some(hint.to_string());
        self
    }

    /// Sets the required flag for the prompt.
    pub fn with_required(&mut self, required: bool) -> &mut Self {
        self.required = required;
        self
    }

    /// Sets the minimum number of selections for the prompt.
    pub fn with_min(&mut self, value: usize) -> &mut Self {
        self.min = value;
        self
    }

    /// Sets the maximum number of selections for the prompt.
    pub fn with_max(&mut self, value: usize) -> &mut Self {
        self.max = value;
        self
    }

    /// Sets the page size for the prompt.
    pub fn with_page_size(&mut self, page_size: usize) -> &mut Self {
        self.page_size = page_size;
        self
    }

    fn values(&mut self) -> Vec<T> {
        self.options
            .iter()
            .filter_map(|option| {
                if option.selected {
                    Some(option.value.clone())
                } else {
                    None
                }
            })
            .collect::<Vec<_>>()
    }

    fn map_options<F>(&mut self, f: F) -> Vec<MultiSelectOption<T>>
    where
        F: Fn(&MultiSelectOption<T>) -> MultiSelectOption<T>,
    {
        self.options.iter().map(f).collect::<Vec<_>>()
    }
}

impl<T: Default + Clone> AsMut<MultiSelect<T>> for MultiSelect<T> {
    fn as_mut(&mut self) -> &mut MultiSelect<T> {
        self
    }
}

impl<T: Default + Clone> Prompt for MultiSelect<T> {
    type Output = Vec<T>;

    fn setup(&mut self) -> Result<(), Error> {
        if self.options.is_empty() {
            return Err(Error::Config("options cannot be empty.".into()));
        }

        if self.min > self.max {
            return Err(Error::Config(format!(
                "min cannot be greater than max (min={}, max={})",
                self.min, self.max
            )));
        }

        Ok(())
    }

    fn handle(&mut self, code: KeyCode, modifiers: KeyModifiers) -> crate::PromptState {
        match (code, modifiers) {
            (KeyCode::Esc, _) | (KeyCode::Char('c'), KeyModifiers::CONTROL) => PromptState::Cancel,
            (KeyCode::Enter, _) => {
                let values = self.values();
                if values.is_empty() && self.required {
                    PromptState::Error(self.formatter.err_required())
                } else if values.len() < self.min {
                    PromptState::Error(self.formatter.err_min(self.min))
                } else if values.len() > self.max {
                    PromptState::Error(self.formatter.err_max(self.max))
                } else {
                    PromptState::Submit
                }
            }
            (KeyCode::Up, _)
            | (KeyCode::Char('k'), _)
            | (KeyCode::Char('p'), KeyModifiers::CONTROL) => {
                self.index = self.index.saturating_sub(1);
                PromptState::Active
            }
            (KeyCode::Down, _)
            | (KeyCode::Char('j'), _)
            | (KeyCode::Char('n'), KeyModifiers::CONTROL) => {
                self.index = std::cmp::min(
                    self.options.len().saturating_sub(1),
                    self.index.saturating_add(1),
                );
                PromptState::Active
            }
            (KeyCode::Char(' '), KeyModifiers::NONE) => {
                let mut option = self.options.get(self.index).unwrap().clone();
                option.selected = !option.selected;
                self.options[self.index] = option;
                PromptState::Active
            }
            (KeyCode::Char('a'), KeyModifiers::NONE) => {
                self.options = self.map_options(|option| {
                    let mut option = option.clone();
                    option.selected = true;
                    option
                });
                PromptState::Active
            }
            (KeyCode::Char('i'), KeyModifiers::NONE) => {
                self.options = self.map_options(|option| {
                    let mut option = option.clone();
                    option.selected = !option.selected;
                    option
                });
                PromptState::Active
            }
            _ => PromptState::Active,
        }
    }

    fn submit(&mut self) -> Self::Output {
        self.values()
    }

    fn render(&mut self, state: &PromptState) -> Result<RenderPayload, String> {
        let payload = RenderPayload::new(self.message.clone(), self.hint.clone(), None);

        match state {
            PromptState::Submit => {
                let raw = self.formatter.submit(
                    self.options
                        .iter()
                        .filter_map(|option| {
                            if option.selected {
                                Some(option.label.clone())
                            } else {
                                None
                            }
                        })
                        .collect::<Vec<_>>(),
                );
                Ok(payload.input(PromptInput::Raw(raw)))
            }

            _ => {
                let page = paginate(self.page_size, &self.options, self.index);
                let options = page
                    .items
                    .iter()
                    .enumerate()
                    .map(|(i, option)| {
                        let active = i == page.cursor;
                        let selected = option.selected;
                        self.formatter.option(
                            self.formatter.option_icon(active, selected),
                            self.formatter
                                .option_label(option.label.clone(), active, selected),
                            self.formatter
                                .option_hint(option.hint.clone(), active, selected),
                            active,
                            selected,
                        )
                    })
                    .collect::<Vec<_>>()
                    .join("\n");

                Ok(payload.body(PromptBody::Raw(options)))
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_prompt;

    macro_rules! options {
        ($count: expr) => {{
            let mut options = Vec::new();
            for i in 1..=$count {
                options.push(MultiSelectOption::new(
                    format!("Value{}", i),
                    format!("value{}", i),
                ));
            }
            options
        }};
    }

    test_prompt!(
        test_hint,
        MultiSelect::new("test message", options!(3)).with_hint("hint message"),
        vec![]
    );

    test_prompt!(
        test_10_items_with_5_page_size,
        MultiSelect::new("test message", options!(10)).with_page_size(5),
        vec![]
    );

    test_prompt!(
        test_option_hint,
        MultiSelect::new(
            "test message",
            vec![
                MultiSelectOption::new("Value1", "value1".to_string()).with_hint("hint1"),
                MultiSelectOption::new("Value2", "value2".to_string()),
                MultiSelectOption::new("Value3", "value3".to_string()).with_hint("hint3"),
            ]
        )
        .with_page_size(5),
        vec![]
    );

    test_prompt!(
        test_move,
        MultiSelect::new("test message", options!(10)).with_page_size(5),
        vec![
            (KeyCode::Char('j'), KeyModifiers::NONE),
            (KeyCode::Char('n'), KeyModifiers::CONTROL),
            (KeyCode::Char('k'), KeyModifiers::NONE),
            (KeyCode::Char('p'), KeyModifiers::CONTROL),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE), // 10
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Up, KeyModifiers::NONE),
            (KeyCode::Up, KeyModifiers::NONE),
            (KeyCode::Up, KeyModifiers::NONE),
            (KeyCode::Up, KeyModifiers::NONE),
            (KeyCode::Up, KeyModifiers::NONE),
            (KeyCode::Up, KeyModifiers::NONE),
            (KeyCode::Up, KeyModifiers::NONE),
            (KeyCode::Up, KeyModifiers::NONE),
            (KeyCode::Up, KeyModifiers::NONE), // 1
            (KeyCode::Up, KeyModifiers::NONE),
        ]
    );

    test_prompt!(
        test_select_2_and_5,
        MultiSelect::new("test message", options!(10)).with_page_size(5),
        vec![
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Char(' '), KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Char(' '), KeyModifiers::NONE),
            (KeyCode::Enter, KeyModifiers::NONE),
        ]
    );

    test_prompt!(
        test_select_all_and_inverse,
        MultiSelect::new("test message", options!(5)).as_mut(),
        vec![
            (KeyCode::Char('a'), KeyModifiers::NONE),
            (KeyCode::Char('i'), KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Char(' '), KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Char(' '), KeyModifiers::NONE),
            (KeyCode::Char('i'), KeyModifiers::NONE),
            (KeyCode::Enter, KeyModifiers::NONE),
        ]
    );

    test_prompt!(
        test_required_error,
        MultiSelect::new("test message", options!(5)).with_required(true),
        vec![(KeyCode::Enter, KeyModifiers::NONE)]
    );

    test_prompt!(
        test_non_required_empty_submit,
        MultiSelect::new("test message", options!(5)).with_required(false),
        vec![(KeyCode::Enter, KeyModifiers::NONE)]
    );

    test_prompt!(
        test_min_error,
        MultiSelect::new("test message", options!(5)).with_min(2),
        vec![
            (KeyCode::Enter, KeyModifiers::NONE),
            (KeyCode::Char(' '), KeyModifiers::NONE),
            (KeyCode::Enter, KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Char(' '), KeyModifiers::NONE),
            (KeyCode::Enter, KeyModifiers::NONE),
        ]
    );

    test_prompt!(
        test_max_error,
        MultiSelect::new("test message", options!(5)).with_max(3),
        vec![
            (KeyCode::Enter, KeyModifiers::NONE),
            (KeyCode::Char(' '), KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Char(' '), KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Char(' '), KeyModifiers::NONE),
            (KeyCode::Down, KeyModifiers::NONE),
            (KeyCode::Char(' '), KeyModifiers::NONE),
            (KeyCode::Enter, KeyModifiers::NONE),
            (KeyCode::Char(' '), KeyModifiers::NONE),
            (KeyCode::Enter, KeyModifiers::NONE),
        ]
    );
}
