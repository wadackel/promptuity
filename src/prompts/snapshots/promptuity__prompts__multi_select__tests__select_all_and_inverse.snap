---
source: src/prompts/multi_select.rs
---
state: Active
message: test message
hint: none
placeholder: none
input(none):
body(raw):
[38;5;8mâ—¯[39m [4mValue1[0m
[38;5;8mâ—¯[39m [38;5;8mValue2[39m
[38;5;8mâ—¯[39m [38;5;8mValue3[39m
[38;5;8mâ—¯[39m [38;5;8mValue4[39m
[38;5;8mâ—¯[39m [38;5;8mValue5[39m
---
state: Active
input(none):
body(raw):
[38;5;10mâ—‰[39m [4mValue1[0m
[38;5;10mâ—‰[39m [38;5;8mValue2[39m
[38;5;10mâ—‰[39m [38;5;8mValue3[39m
[38;5;10mâ—‰[39m [38;5;8mValue4[39m
[38;5;10mâ—‰[39m [38;5;8mValue5[39m
---
state: Active
input(none):
body(raw):
[38;5;8mâ—¯[39m [4mValue1[0m
[38;5;8mâ—¯[39m [38;5;8mValue2[39m
[38;5;8mâ—¯[39m [38;5;8mValue3[39m
[38;5;8mâ—¯[39m [38;5;8mValue4[39m
[38;5;8mâ—¯[39m [38;5;8mValue5[39m
---
state: Active
input(none):
body(raw):
[38;5;8mâ—¯[39m [38;5;8mValue1[39m
[38;5;8mâ—¯[39m [4mValue2[0m
[38;5;8mâ—¯[39m [38;5;8mValue3[39m
[38;5;8mâ—¯[39m [38;5;8mValue4[39m
[38;5;8mâ—¯[39m [38;5;8mValue5[39m
---
state: Active
input(none):
body(raw):
[38;5;8mâ—¯[39m [38;5;8mValue1[39m
[38;5;10mâ—‰[39m [4mValue2[0m
[38;5;8mâ—¯[39m [38;5;8mValue3[39m
[38;5;8mâ—¯[39m [38;5;8mValue4[39m
[38;5;8mâ—¯[39m [38;5;8mValue5[39m
---
state: Active
input(none):
body(raw):
[38;5;8mâ—¯[39m [38;5;8mValue1[39m
[38;5;10mâ—‰[39m [38;5;8mValue2[39m
[38;5;8mâ—¯[39m [4mValue3[0m
[38;5;8mâ—¯[39m [38;5;8mValue4[39m
[38;5;8mâ—¯[39m [38;5;8mValue5[39m
---
state: Active
input(none):
body(raw):
[38;5;8mâ—¯[39m [38;5;8mValue1[39m
[38;5;10mâ—‰[39m [38;5;8mValue2[39m
[38;5;10mâ—‰[39m [4mValue3[0m
[38;5;8mâ—¯[39m [38;5;8mValue4[39m
[38;5;8mâ—¯[39m [38;5;8mValue5[39m
---
state: Active
input(none):
body(raw):
[38;5;10mâ—‰[39m [38;5;8mValue1[39m
[38;5;8mâ—¯[39m [38;5;8mValue2[39m
[38;5;8mâ—¯[39m [4mValue3[0m
[38;5;10mâ—‰[39m [38;5;8mValue4[39m
[38;5;10mâ—‰[39m [38;5;8mValue5[39m
---
state: Submit
input(raw): Value1, Value4, Value5
body(none):
